### DISCUSSION TITLE
Vorstellung von rr2mqtt - Nutzen des Adapters nur per lokalen MQTT

### DISCUSSION THREAD

**Discussion by csowada** [copystring/ioBroker.roborock#704] [Ideas]
Moin und schöne Grüße aus Hamburg.

Ich wollte mein kleines Mini Projekt auf Basis von ioBroker.roborock vorstellen. Aktuell ist es nur eine Idee bzw. PoC um diesen sehr ausgereiften Adapter auch für andere Heim-Automatisierungen zu nutzen. Ich nutze z.B. openHAB und auch sehr viel Zigbee2MQTT. Daher die Idee den Adapter mit wenig Aufwand auch nur auf MQTT Basis zu betreiben.

Ich hatte mit @copystring bereits in einem Bug die Diskussion angefangen. Aber ich wollte den Bug nicht kapern.
Aber der Vorschlag von @copystring eine gemeinsame Bibliothek zu extrahieren finde ich sehr interessant. Auch wenn ich es mir mit meinem Ansatz sehr einfach gemacht habe, auf eine sehr coole Lösung aufzusetzen. 

Daher starte ich heir einfach mal eine Diskussion.

Was macht rr2mqtt eigentlich. Eigentlich ganz einfach, der ioBroker Adapter wird durch ein Fake Adapter mit der Basisimplementierung ausgetauscht. Dadurch wird ioBroker.roborock schon Standalone lauffähig. Gute Arbeit! Ich musste daher nur eine ganz dünnen Adapter [rr2mqtt-adapter.js](https://github.com/csowada/rr2mqtt/blob/main/rr2mqtt-adapter.js) bauen und schon konnte die States und Objects einfach an den lokalen MQTT Broker gesendet werden. Damit kann dann jede Heim-Automatisierungen auf diesen tollen Adapter zugreifen. Das ganze dann noch in einen Docker Container verpackt und jeder kann mit wenig Aufwand starten.

Ich hoffen natürlich, dass ist auch in @copystring Sinne. Grundsätzlich bringe ich meine Ideen, Bugfixes etc. so oder so in dieses Projekt ein.

[Git Projekt](https://github.com/csowada/rr2mqtt/tree/main)

Schöne Grüße

Christian

**Comment by copystring** [comment]
Hallo @csowada,

erstmal danke für das Eröffnen der Diskussion. Sowas regt neue Ideen an und kann sehr konstruktiv sein!

An sich habe ich nichts dagegen, wenn du rr2mqtt basierend auf meinem Code entwickelst.

- Wusstest du, dass die Roborock App ohnehin über MQTT kommuniziert und somit auch dieser Adapter, wenn keine TCP-Verbindung zum Saugroboter aufgebaut werden kann?
- rr2mqtt wird immer funktional hinterherhinken, wenn dieser lediglich basierend auf meinem Adapter ist. Es kommen ständig neue Saugroboter raus, welche neue Parameter haben. Diese würdest du dann immer nur basierend auf meinem Adapter implementieren? Genau dort würde ich den großen Vorteil einer entsprechenden Bibliothek sehen.

Meiner persönlichen Meinung nach, wäre es genau die vergeudete Zeit, welche besser in eine allgemeine Bibliothek gesteckt werden könnte, welche man gemeinsam pflegen könnte.

rr2mqtt wird ein fundamentales Problem haben, wenn rr2mqtt kein UI für den Anwender bereitstellt, um z. B. Koordinaten für die Zonenreinigung und Goto-Befehl herauszufinden. Welche Pläne hast du dafür?

Ich freue mich auf dein Feedback und würde mich freuen, wenn hieraus vielleicht eine allgemeine Bibliothek entsteht, welche auch für andere Projekte verwendet werden kann.

**Comment by csowada** [comment]
Moin @copystring,

ich finde die Idee mit einer eigenen Bibliothek wirklich supper. Ich würde mein PoC Projekt aber erstmal auf ein stabileres Level bekommen wollen. Dafür brauche ich noch ein paar Tage. Die Lösung läuft bei mir gerade im Probebetrieb.

Schwebt dir schon etwas vor, wie man den aktuellen Code aufteilen könnte? Z.B. Core, Webserver und ioBroker Adapter? Man kann in meinem Fake Adapter schon erkennen, das die Abhängigkeit gar nicht groß ist. Als Bibliothek wäre langfristig auch Typescript interessant.
